clients/src/main/java/org/apache/kafka/common/telemetry/internals/LastValueTracker.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.telemetry.internals;
+
+import java.time.Instant;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A LastValueTracker uses a ConcurrentMap to maintain historic values for a given key, and return
+ * a previous value and an Instant for that value.
+ *
+ * @param <T> The type of the value.
+ */
+public class LastValueTracker<T> {
+    private final ConcurrentMap<MetricKey, AtomicReference<InstantAndValue<T>>> counters = new ConcurrentHashMap<>();
+
+    /**
+     * Return the last instant/value for the given MetricKey, or Optional.empty if there isn't one.
+     *
+     * @param metricKey the key for which to calculate a getAndSet.
+     * @param now the timestamp for the new value.
+     * @param value the current value.
+     * @return the timestamp of the previous entry and its value. If there
+     *     isn't a previous entry, then this method returns {@link Optional#empty()}
+     */
+    public Optional<InstantAndValue<T>> getAndSet(MetricKey metricKey, Instant now, T value) {
+        InstantAndValue<T> instantAndValue = new InstantAndValue<>(now, value);
+        AtomicReference<InstantAndValue<T>> valueOrNull = counters
```

Comment:
Not sure why we need to wrap `instantAndValue` in an `AtomicReference`. Can you elaborate?

Reply:
To make the operations thread safe as MetricsCollector can invoke the method concurrently.

Reply:
Still not sure if I can follow -- I don't see any code that would return the value of the `counters` map? And the put is protected by the concurrent map.

Reply:
@mjsax Ahh I see what you mean, thanks. Yes agree, if we want to implement just `getAndSet` which updates the latest value and returns last held value then invoking `counters.put(instantAndValue)` would have updated the result atomically while providing last held value.

Tagging @xvrl if he has some suggestion here as I think he originally reviewed the code for this class. If there isn't anything we are missing then I ll remove AtomicReference and replace `putIfAbsent` to `put, while returning the value returned from `put` as method return argument.

Reply:
Agree, I think in this case the put should be sufficient.

Reply:
I have removed AtomicReference, done.